File: src/sandbox.cpp
#include "sandbox.hpp"

#include <cerrno>
#include <cstring>
#include <stdexcept>
#include <system_error>

#ifdef __linux__
#  include <grp.h>
#  include <pwd.h>
#  include <unistd.h>
#endif

namespace sandbox {

UserInfo lookup_user(const std::string& username) {
#ifdef __linux__
    errno = 0;
    const struct passwd* pw = ::getpwnam(username.c_str());
    if (!pw)
        throw std::runtime_error(errno
            ? "getpwnam(" + username + "): " + std::strerror(errno)
            : "user not found: " + username);
    return {pw->pw_uid, pw->pw_gid, username};
#else
    (void)username;
    throw std::runtime_error("sandbox::lookup_user: not supported on this platform");
#endif
}

void chown_jail(const std::filesystem::path& dir, const UserInfo& user) {
#ifdef __linux__
    if (::geteuid() == 0 &&
        ::chown(dir.c_str(), user.uid, user.gid) != 0)
        throw std::system_error(errno, std::generic_category(),
                                "chown(" + dir.string() + ")");
#else
    (void)dir; (void)user;
#endif
}

void enter_chroot(const std::filesystem::path& dir) {
#ifdef __linux__
    if (::chroot(dir.c_str()) != 0)
        throw std::system_error(errno, std::generic_category(),
                                "chroot(" + dir.string() + ")");
    if (::chdir("/") != 0)
        throw std::system_error(errno, std::generic_category(),
                                "chdir(/) after chroot");
#else
    (void)dir;
    throw std::runtime_error("sandbox::enter_chroot: not supported on this platform");
#endif
}

void drop_privileges(const UserInfo& user) {
#ifdef __linux__
    if (::setgroups(0, nullptr) != 0)
        throw std::system_error(errno, std::generic_category(), "setgroups");
    if (::setgid(user.gid) != 0)
        throw std::system_error(errno, std::generic_category(), "setgid");
    if (::setuid(user.uid) != 0)
        throw std::system_error(errno, std::generic_category(), "setuid");
    // Sanity check: after a successful drop, setuid(0) must fail.
    if (::setuid(0) == 0)
        throw std::runtime_error(
            "sandbox: privilege drop verification failed - "
            "process can still re-acquire UID 0");
#else
    (void)user;
    throw std::runtime_error("sandbox::drop_privileges: not supported on this platform");
#endif
}

} // namespace sandbox

File: src/sandbox.rs
use std::path::Path;

use anyhow::{Context, Result, bail};
use nix::{
    sys::resource::{Resource, getrlimit, setrlimit},
    unistd::{
        Uid, User, chdir, chown, chroot, getegid, geteuid, getgid, getuid, setgid, setgroups,
        setuid,
    },
};

pub fn apply_resource_limits() -> Result<()> {
    setrlimit(Resource::RLIMIT_CORE, 0, 0).context("disabling core dumps")?;
    let (_, hard_limit) =
        getrlimit(Resource::RLIMIT_NOFILE).context("reading the descriptor limit")?;
    let descriptor_limit = hard_limit.min(4096);
    setrlimit(Resource::RLIMIT_NOFILE, descriptor_limit, descriptor_limit)
        .context("limiting open descriptors")
}

pub fn resolve_user(name: Option<&str>) -> Result<Option<User>> {
    let Some(name) = name else {
        return Ok(None);
    };
    let user = User::from_name(name)
        .context("looking up --user")?
        .ok_or_else(|| anyhow::anyhow!("system user '{name}' does not exist"))?;
    if user.uid.is_root() {
        bail!("--user must name an unprivileged account");
    }
    Ok(Some(user))
}

pub fn require_root(sandbox: bool, user: Option<&User>) -> Result<()> {
    if (sandbox || user.is_some()) && !geteuid().is_root() {
        bail!("--sandbox and --user require root privileges");
    }
    Ok(())
}

pub fn warn_if_root(user: Option<&User>) {
    if geteuid().is_root() && user.is_none() {
        tracing::warn!("running as root; use --user to reduce process privileges");
    }
}

pub fn assign_data_directory(data_dir: &Path, user: &User) -> Result<()> {
    chown(data_dir, Some(user.uid), Some(user.gid)).context("changing data directory owner")
}

pub fn enter(data_dir: &Path) -> Result<()> {
    chroot(data_dir).with_context(|| format!("entering chroot at {}", data_dir.display()))?;
    chdir("/").context("changing to the chroot root")?;
    Ok(())
}

pub fn drop_privileges(user: &User) -> Result<()> {
    setgroups(&[]).context("clearing supplementary groups")?;
    setgid(user.gid).context("setting process group")?;
    setuid(user.uid).context("setting process user")?;

    if getuid() != user.uid
        || geteuid() != user.uid
        || getgid() != user.gid
        || getegid() != user.gid
    {
        bail!("privilege-drop verification failed");
    }
    if setuid(Uid::from_raw(0)).is_ok() {
        bail!("privilege-drop verification failed: root can be reacquired");
    }
    tracing::info!(user = %user.name, uid = user.uid.as_raw(), gid = user.gid.as_raw(), "privileges dropped and verified");
    Ok(())
}
